crates/runtime/src/dataconnector/sharepoint.rs

```diff
@@ -0,0 +1,243 @@
+        .with_scope(scope)
+        .with_tenant(tenant_id)
+        .with_state(state)
+        .url()
+}
+
+static NEEDED_SCOPE: [&str; 2] = ["User.Read", "Files.Read"];
+static REDIRECT_URL: &str = "http://localhost:8090/redirect/sharepoint";
+
+impl Sharepoint {
+    fn new(params: &Parameters) -> Result<Self> {
+        let client_id = params
+            .get("client_id")
+            .expose()
+            .ok_or_else(|p| MissingParameterSnafu { parameter: p.0 }.build())?;
+        let client_secret = params
+            .get("client_secret")
+            .expose()
+            .ok_or_else(|p| MissingParameterSnafu { parameter: p.0 }.build())?;
+        let tenant_id = params
+            .get("tenant_id")
+            .expose()
+            .ok_or_else(|p| MissingParameterSnafu { parameter: p.0 }.build())?;
+        let redirect_url_str = params
+            .get("redirect_url")
+            .expose()
+            .ok()
+            .unwrap_or(REDIRECT_URL);
+        let redirect_url = Url::parse(redirect_url_str).map_err(|_| {
+            InvalidRedirectUrlSnafu {
+                url: redirect_url_str.to_string(),
+            }
+            .build()
+        })?;
+
+        let graph_client =
+            if let Some(authorization_code) = params.get("authorization_code").expose().ok() {
+                Ok(auth_code_grant_secret(
+                    authorization_code,
+                    client_id,
+                    client_secret,
+                    NEEDED_SCOPE.map(ToString::to_string).to_vec(),
+                    redirect_url,
+                ))
+            } else {
+                let authorization_url = authorization_sign_in_url(
+                    client_id,
+                    "",
+                    tenant_id,
+                    redirect_url,
+                    NEEDED_SCOPE.map(ToString::to_string).to_vec(),
+                )
+                .context(AuthorizationUrlSnafu)?;
+                println!("Please sign in to the following URL: {authorization_url}");
+
+                Err(AuthorizationCodeSnafu {}.build())?
+            }?;
+
+        Ok(Sharepoint {
+            client: Arc::new(graph_client),
+        })
+    }
+}
+
+#[derive(Default, Copy, Clone)]
+pub struct SharepointFactory {}
+
+impl SharepointFactory {
+    #[must_use]
+    pub fn new() -> Self {
+        Self {}
+    }
+
+    #[must_use]
+    pub fn new_arc() -> Arc<dyn DataConnectorFactory> {
+        Arc::new(Self {}) as Arc<dyn DataConnectorFactory>
+    }
+}
+
+const PARAMETERS: &[ParameterSpec] = &[
+    ParameterSpec::runtime("client_id")
+        .secret()
+        .description("")
+        .required(),
+    ParameterSpec::runtime("client_secret")
+        .secret()
+        .description("")
+        .required(),
+    ParameterSpec::runtime("tenant_id")
+        .secret()
+        .description("")
+        .required(),
+    ParameterSpec::runtime("redirect_url").description(""),
+    ParameterSpec::runtime("authorization_code")
```

Comment:
`authorization_code` is not marked as required to provide better UX via constructing sign in URL? https://github.com/spiceai/spiceai/pull/2294/files#diff-dc08f6b1dfbb94cffb5c12d2fee5b414952d3583ad3da03d442985ff1dc04e5fR136

Reply:
not needed anymore, hopefully.